### PR TITLE
Extension registry client API

### LINKED ISSUE #12422: review and enable registry mechanism

### PR DESCRIPTION
**Pull request by aloubyansky**
This PR:
* switches the devtools from the legacy QuarkusPlatformDescriptor based API to the new ExtensionCatalog API;
* introduces the registry client API and switches Maven and Gradle plaugins and the CLI to use the registry client to resolve the catalog;
* changes the platform descriptor JSON format.

Until the default community registry service is available I implemented a fallback to resolving the platform descriptors directly from the BOM coordinates.

Fixes #12422

### CODE REVIEW

File: independent-projects/tools/devtools-common/src/main/java/io/quarkus/devtools/codestarts/quarkus/QuarkusCodestartCatalog.java [hunk 1]
```diff
@@ -62,29 +61,12 @@ private QuarkusCodestartCatalog(Collection<Codestart> codestarts,
...
+    public static QuarkusCodestartCatalog fromExtensionsCatalog(ExtensionCatalog catalog, ResourceLoader resourceLoader)
             throws IOException {
-        final CodestartPathLoader pathLoader = platformPathLoader(platformDescriptor);
+        final CodestartPathLoader pathLoader = platformPathLoader(resourceLoader);
         final Collection<Codestart> codestarts = CodestartCatalogLoader.loadCodestarts(pathLoader, QUARKUS_CODESTARTS_DIR);
-        final Map<String, Extension> extensionsMapping = buildExtensionsMapping(platformDescriptor.getExtensions());
-        return new QuarkusCodestartCatalog(codestarts, extensionsMapping);
-    }
-
-    public static QuarkusCodestartCatalog fromQuarkusPlatformDescriptorAndDirectories(
```
Comment (ia3andy) [inline independent-projects/tools/devtools-common/src/main/java/io/quarkus/devtools/codestarts/quarkus/QuarkusCodestartCatalog.java:73]: @aloubyansky this is used from outside..

Comment (ia3andy) [reply independent-projects/tools/devtools-common/src/main/java/io/quarkus/devtools/codestarts/quarkus/QuarkusCodestartCatalog.java:73]: https://github.com/quarkusio/jbang-catalog/blob/master/src/QuarkusExampleCodestartDev.java#L140

**Review by ia3andy — CHANGES_REQUESTED** [review]

File: independent-projects/tools/devtools-common/src/main/java/io/quarkus/devtools/codestarts/quarkus/QuarkusCodestartCatalog.java (hunk 1 above)
Comment (aloubyansky) [reply independent-projects/tools/devtools-common/src/main/java/io/quarkus/devtools/codestarts/quarkus/QuarkusCodestartCatalog.java:73]: Well, QuarkusPlatformDescriptor does not exist any more. So the caller will have to call new methods.

Comment (ia3andy) [reply independent-projects/tools/devtools-common/src/main/java/io/quarkus/devtools/codestarts/quarkus/QuarkusCodestartCatalog.java:73]: @aloubyansky yeah we just need to have the `ExtensionCatalog catalog, ResourceLoader resourceLoader, Collection<Path> directories` possibility

**Comment by gsmet** [comment]
> Until the default community registry service is available I implemented a fallback to resolving the platform descriptors directly from the BOM coordinates.

Small comment on this: I think you will need to keep the fallback around. It's not obvious people will be able to access the registry in corporate environments (a lot of corporate environments have exceptions for Maven Central from what I saw these past months in user comments).

File: independent-projects/tools/devtools-common/src/main/java/io/quarkus/devtools/codestarts/quarkus/QuarkusCodestartCatalog.java (hunk 1 above)
Comment (aloubyansky) [reply independent-projects/tools/devtools-common/src/main/java/io/quarkus/devtools/codestarts/quarkus/QuarkusCodestartCatalog.java:73]: It should be back now.

**Review by ia3andy — DISMISSED** [review]

**Comment by aloubyansky** [comment]
> > Until the default community registry service is available I implemented a fallback to resolving the platform descriptors directly from the BOM coordinates.
> 
> Small comment on this: I think you will need to keep the fallback around. It's not obvious people will be able to access the registry in corporate environments (a lot of corporate environments have exceptions for Maven Central from what I saw these past months in user comments).

Interestingly, one of the reasons to use the Maven backend/interface was to make it easily/easier available in corporate environments. If our tools will end being limited to the platform extensions in (many) corp.env. or be painful to configure we should re-consider the approach. One way to make it available OOTB would be to publish the non-platform extension catalogs to the Maven Central repo.

File: independent-projects/tools/registry-client/src/main/java/io/quarkus/registry/catalog/json/JsonCatalogMapperHelper.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.Reader;
+import java.io.Writer;
+import java.nio.file.Files;
+import java.nio.file.Path;
+
+public class JsonCatalogMapperHelper {
+
+    private static ObjectMapper mapper;
```
Comment (gastaldi) [inline independent-projects/tools/registry-client/src/main/java/io/quarkus/registry/catalog/json/JsonCatalogMapperHelper.java:20]: Can this mapper be set externally somehow?

File: independent-projects/tools/registry-client/src/main/java/io/quarkus/registry/config/json/RegistriesConfigMapperHelper.java [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+        if (Files.exists(p)) {
+            Files.delete(p);
+        }
+        serialize(resolver, p);
+        return p;
+    }
+
+    private static void logDeserialized(Path p) throws IOException {
+        JsonRegistriesConfig resolver = deserialize(p, JsonRegistriesConfig.class);
+        System.out.println("QER RESOLVER:");
```
Comment (gastaldi) [inline independent-projects/tools/registry-client/src/main/java/io/quarkus/registry/config/json/RegistriesConfigMapperHelper.java:131]: Remove the `System.out` here (use logger?)

File: independent-projects/tools/registry-client/src/main/java/io/quarkus/registry/config/json/RegistriesConfigMapperHelper.java [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+        resolver.getRegistries().forEach(qer -> {
+            System.out.println(" - " + qer);
+        });
+    }
+
+    private static void logLoaded(Path p) throws IOException {
+        RegistriesConfig resolver = RegistriesConfigLocator.load(p);
+        System.out.println("QER RESOLVER:");
```
Comment (gastaldi) [inline independent-projects/tools/registry-client/src/main/java/io/quarkus/registry/config/json/RegistriesConfigMapperHelper.java:139]: Remove the `System.out` here (use logger?)

File: independent-projects/tools/devtools-common/pom.xml [hunk 5]
```diff
@@ -28,7 +28,7 @@
     <dependencies>
```
Comment (gastaldi) [inline independent-projects/tools/devtools-common/pom.xml:28]: Remove this dependency:
```xml
        <dependency>
            <groupId>org.immutables</groupId>
            <artifactId>value</artifactId>
            <version>2.8.8</version>
            <scope>provided</scope>
        </dependency>
```

**Review by gastaldi — CHANGES_REQUESTED** [review]
I reviewed the changes on each of the 277 files impacted by this PR and they look decent. I request a couple of small changes though.

I have one suggestion: it would be nice to have a `quarkus-devtools-registry-spi` module containing the SPI classes required to be implemented in the server side of a Registry (eg. the registry.quarkus.io app)

**Comment by aloubyansky** [comment]
> I have one suggestion: it would be nice to have a quarkus-devtools-registry-spi module containing the SPI classes required to be implemented in the server side of a Registry (eg. the registry.quarkus.io app)

This PR doesn't really include anything that should be implemented on the server side. So, the module would be empty.

File: independent-projects/tools/registry-client/src/main/java/io/quarkus/registry/catalog/json/JsonCatalogMapperHelper.java (hunk 2 above)
Comment (aloubyansky) [reply independent-projects/tools/registry-client/src/main/java/io/quarkus/registry/catalog/json/JsonCatalogMapperHelper.java:20]: I added methods to pass in a custom mapper. Would that work?

File: independent-projects/tools/registry-client/src/main/java/io/quarkus/registry/config/json/RegistriesConfigMapperHelper.java (hunk 3 above)
Comment (aloubyansky) [reply independent-projects/tools/registry-client/src/main/java/io/quarkus/registry/config/json/RegistriesConfigMapperHelper.java]: It's a left over, removed.

File: independent-projects/tools/registry-client/src/main/java/io/quarkus/registry/config/json/RegistriesConfigMapperHelper.java (hunk 4 above)
Comment (aloubyansky) [reply independent-projects/tools/registry-client/src/main/java/io/quarkus/registry/config/json/RegistriesConfigMapperHelper.java]: Removed.

File: independent-projects/tools/devtools-common/pom.xml (hunk 5 above)
Comment (aloubyansky) [reply independent-projects/tools/devtools-common/pom.xml:28]: Not sure why I had it, removed.

File: independent-projects/tools/registry-client/src/main/java/io/quarkus/registry/catalog/json/JsonCatalogMapperHelper.java (hunk 2 above)
Comment (gastaldi) [reply independent-projects/tools/registry-client/src/main/java/io/quarkus/registry/catalog/json/JsonCatalogMapperHelper.java:20]: Yes, I think that would help. I was mostly thinking of reusing the `ObjectMapper` produced by the `quarkus-jackson` extension

**Comment by gastaldi** [comment]
Rebased since it introduced conflicts in https://github.com/quarkusio/quarkus/commit/0d0bd91aa0bf4b1d80ebe66aa0b08a05aaca5cf4

**Review by gastaldi — APPROVED** [review]
LGTM

**Pull request merged**